### PR TITLE
Support additional salvage debug info opportunities

### PR DESCRIPTION
**Pull request by adrian-prantl**
This picks up a PR from @mshockwave with some additional bugfixes on top.

### CODE REVIEW

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

File: lib/SIL/Parser/ParseSIL.cpp [hunk 1]
```diff
@@ -1690,6 +1694,23 @@ bool SILParser::parseSILDebugInfoExpression(SILDebugInfoExpression &DIExpr) {
...
         }
+        case SILDIExprElement::ConstIntKind: {
+          bool IsNegative = false;
+          if (P.Tok.is(tok::oper_prefix) && P.Tok.getRawText() == "-") {
+            P.consumeToken();
+            IsNegative = true;
+          }
+          int64_t Val;
+          // FIXME: Use the correct diagnostic
+          if (parseInteger(Val, diag::sil_invalid_scope_slot))
```
Comment (mshockwave) [inline lib/SIL/Parser/ParseSIL.cpp:1705]: Can you create a new diagnostic message for this? (something like "expecting expecting an integer")

File: lib/IRGen/IRGenSIL.cpp [hunk 2]
```diff
@@ -4731,11 +4731,18 @@ void IRGenSILFunction::visitDebugValueInst(DebugValueInst *i) {
...
+  if (auto MaybeSILTy = VarInfo->Type) {
     // If there is auxiliary type info, use it
     SILTy = *MaybeSILTy;
-  else
+  } else {
     SILTy = SILVal->getType();
+    if (auto Expr = VarInfo->DIExpr)
+      if (auto n = Expr.getNumElements())
+        if (Expr.getElement(n - 1).getAsOperator() ==
+            SILDIExprOperator::Fragment)
```
Comment (mshockwave) [inline lib/IRGen/IRGenSIL.cpp:4743]: I think `SILDebugInfoExpression::hasFragment` can be used here

File: lib/IRGen/IRGenSIL.cpp [hunk 3]
```diff
@@ -5110,18 +5118,27 @@ void IRGenSILFunction::emitDebugInfoForAllocStack(AllocStackInst *i,
...
+  if (auto MaybeSILTy = VarInfo->Type) {
     // If there is auxiliary type info, use it
     SILTy = *MaybeSILTy;
-  else
+  } else {
     SILTy = i->getType();
+    if (auto Expr = VarInfo->DIExpr)
+      if (auto n = Expr.getNumElements())
+        if (Expr.getElement(n - 1).getAsOperator() ==
+            SILDIExprOperator::Fragment)
```
Comment (mshockwave) [inline lib/IRGen/IRGenSIL.cpp:5130]: ditto `SILDebugInfoExpression::hasFragment`

**Review by mshockwave — COMMENTED** [review]
Thank you Adrian for wrapping up this patch!
Just want to make sure I understand correctly about this patch: If we're dealing with a debug_value that has a fragment, the debug variable size will essentially be zero. But then I'm wondering: Will LLVM be happy about a `!DIVariable` that has zero size?

File: lib/IRGen/IRGenSIL.cpp (hunk 2 above)
Comment (adrian-prantl) [reply lib/IRGen/IRGenSIL.cpp:4743]: Much better!

**Comment by adrian-prantl** [comment]
@swift-ci test

**Pull request merged**